geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedDataStoreViewFactory.scala

```diff
@@ -0,0 +1,113 @@
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.index.view
+
+import java.awt.RenderingHints
+
+import com.typesafe.config._
+import org.geotools.data.DataAccessFactory.Param
+import org.geotools.data.{DataStore, DataStoreFactorySpi, DataStoreFinder}
+import org.locationtech.geomesa.index.geotools.GeoMesaDataStoreFactory.{GeoMesaDataStoreInfo, NamespaceParams}
+import org.locationtech.geomesa.utils.classpath.ServiceLoader
+import org.locationtech.geomesa.utils.geotools.GeoMesaParam
+
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Data store factory for routed views
+  */
+class RoutedDataStoreViewFactory extends DataStoreFactorySpi {
+
+  import RoutedDataStoreViewFactory._
+
+  import scala.collection.JavaConverters._
+
+  override def canProcess(params: java.util.Map[String, java.io.Serializable]): Boolean =
+    RoutedDataStoreViewFactory.canProcess(params)
+
+  override def createDataStore(params: java.util.Map[String, java.io.Serializable]): DataStore =
+    createNewDataStore(params)
+
+  override def createNewDataStore(params: java.util.Map[String, java.io.Serializable]): DataStore = {
+    val config = ConfigFactory.parseString(ConfigParam.lookup(params))
+    val configs = if (config.hasPath("stores")) { config.getConfigList("stores").asScala } else { Seq.empty }
+    if (configs.isEmpty) {
+      throw new IllegalArgumentException("No 'stores' element defined in configuration")
+    }
+
+    val namespace = NamespaceParam.lookupOpt(params)
+    val nsConfig = namespace.map(ConfigValueFactory.fromAnyRef)
+
+    val stores = Seq.newBuilder[(DataStore, java.util.Map[String, AnyRef])]
+    stores.sizeHint(configs.length)
+
+    try {
+      configs.foreach { config =>
+        lazy val error = new IllegalArgumentException(s"Could not load store using configuration:\n" +
+            config.root().render(ConfigRenderOptions.concise().setFormatted(true)))
+        // inject the namespace into the underlying stores
+        val storeParams = nsConfig.map(config.withValue(NamespaceParam.key, _)).getOrElse(config).root().unwrapped()
+        Try(DataStoreFinder.getDataStore(storeParams)) match {
+          case Success(null)  => throw error
+          case Success(store) => stores += store -> storeParams
+          case Failure(e)     => throw error.initCause(e)
+        }
+      }
+    } catch {
+      case NonFatal(e) => stores.result.foreach(_._1.dispose()); throw e
+    }
+    val storesAndConfigs = stores.result
+
+    val router = {
+      val impl = RouterParam.lookup(params)
+      ServiceLoader.load[RouteSelector]().find(_.getClass.getName == impl).getOrElse {
+        throw new IllegalArgumentException(s"Could not load route selector '$impl'. Available implementations: " +
```

Comment:
If we die here, should we clean up the DataStores which have been instantiated?

Reply:
good catch, fixed